File: app/src/main/java/com/lzm/lightLive/act/IntroActivity.java
package com.lzm.lightLive.act;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModel;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;
import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Pair;
import android.util.SparseArray;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.FrameLayout;

import com.airbnb.lottie.LottieDrawable;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.android.material.transition.platform.MaterialContainerTransformSharedElementCallback;
import com.lzm.lib_base.BaseBindFragmentActivity;
import com.lzm.lib_base.BaseFreshFragment;
import com.lzm.lightLive.R;
import com.lzm.lightLive.adapter.FragmentAdapter;
import com.lzm.lightLive.databinding.ActivityIntroBinding;
import com.lzm.lightLive.frag.AllCategoryFragment;
import com.lzm.lightLive.frag.HomeFragment;
import com.lzm.lightLive.frag.SettingFragment;
import com.lzm.lightLive.frag.AllSubscribedFragment;
import com.lzm.lightLive.util.UiTools;
import java.util.ArrayList;
import java.util.List;

public class IntroActivity extends BaseBindFragmentActivity<ActivityIntroBinding, ViewModel>
        implements BottomNavigationView.OnItemSelectedListener {

    private static final String TAG = "IntroActivity";

    private MenuItem itemHome;
    private MenuItem itemTag;
    private MenuItem itemUser;
    private MenuItem itemSearch;

    private final List<MenuItem> menuItems = new ArrayList<>();

    private FragmentAdapter fragmentPager;

    private int appBarSlideOffset = -1;

    private float percent = 0f;

    @Override
    protected int getLayoutResId() {
        return R.layout.activity_intro;
    }

    @Override
    protected ViewModel getViewModel() {
        return null;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                | View.SYSTEM_UI_FLAG_LAYOUT_STABLE);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        getWindow().setStatusBarColor(Color.TRANSPARENT);
        getWindow().getDecorView().setSystemUiVisibility(View.GONE);
        Window window = getWindow();
        window.requestFeature(Window.FEATURE_ACTIVITY_TRANSITIONS);
        setExitSharedElementCallback(new MaterialContainerTransformSharedElementCallback());
        window.setSharedElementsUseOverlay(false);

        getWindow().setStatusBarColor(Color.TRANSPARENT);
        super.onCreate(savedInstanceState);

        initBottomNav();
        initFragment();
        mBind.appBarLayout.setAlpha(1);
        mBind.appBarLayout.post(this::initAppBarListener);
        showFragment(0);
    }

    private void initFragment() {

        Pair<Integer, Fragment> home = new Pair<>(0, new HomeFragment());
        Pair<Integer, Fragment> cate = new Pair<>(1, new AllCategoryFragment());
        Pair<Integer, Fragment> sub = new Pair<>(2, new AllSubscribedFragment());
        Pair<Integer, Fragment> set = new Pair<>(3, new SettingFragment());

        fragmentPager = new FragmentAdapter(this, home, cate, sub, set);
        mBind.content.setOffscreenPageLimit(2);
        mBind.content.setUserInputEnabled(true);
        mBind.content.setAdapter(fragmentPager);
        mBind.content.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                mBind.bottomNavigation.setSelectedItemId(menuItems.get(position).getItemId());
                super.onPageSelected(position);
            }
        });
    }

    private void initAppBarListener() {
        ViewGroup.LayoutParams barLp = mBind.actionBar.getLayoutParams();
        FrameLayout.LayoutParams tvTitleLp = (FrameLayout.LayoutParams) mBind.tvTitle.getLayoutParams();

        int barHeight = mBind.actionBar.getHeight();
        int tvTitleTopMargin = tvTitleLp.topMargin;
        int tvTitleLeftMargin = tvTitleLp.leftMargin;    //UiTools.dp2px(this, 30);

        float textSize = UiTools.pxToDp(this, mBind.tvTitle.getTextSize());

        mBind.appBarLayout.addOnOffsetChangedListener((appBarLayout, verticalOffset) -> {
            if (verticalOffset == appBarSlideOffset) {
                return;
            }
            appBarSlideOffset = verticalOffset;
            percent = Math.abs(verticalOffset * 1.0f) / mBind.anchorView.getTop();
            if (percent != 0f) {
                barLp.height = (int) (barHeight - barHeight * (0.7 * percent));
                tvTitleLp.topMargin = (int) (tvTitleTopMargin - tvTitleTopMargin * (0.5 * percent));
                tvTitleLp.leftMargin = (int) (tvTitleLeftMargin + (tvTitleLeftMargin * percent));

                runOnUiThread(() -> {
                    mBind.actionBar.setLayoutParams(barLp);
                    mBind.tvTitle.setLayoutParams(tvTitleLp);
                    mBind.tvTitle.setTextSize((float) (textSize - (textSize * (0.3 * percent))));
                });
            }
            if (Math.abs(verticalOffset) >= mBind.anchorView.getTop()) {
//                mBind.tvTitle.setTextColor(Color.WHITE);
//                mBind.ivTitle.setTextColor(Color.WHITE);
                mBind.actionBarBlur.setVisibility(View.VISIBLE);
                UiTools.setStatusBar(this, true);
            } else {
                mBind.actionBarBlur.setVisibility(View.VISIBLE);
//                mBind.tvTitle.setTextColor(Color.WHITE);
//                DrawableCompat.setTint(mBind.ivTitle.getDrawable(), getResources().getColor(R.color.black));
                mBind.actionBarBlur.setVisibility(View.GONE);
                UiTools.setStatusBar(this, false);
            }
        });
    }

    private void initBottomNav() {
        menuItems.add(itemHome);
        menuItems.add(itemTag);
        menuItems.add(itemUser);
        menuItems.add(itemSearch);
        for (MenuItem item : menuItems) {
            @StringRes int string = -1;
            @DrawableRes int drawable = -1;
            int position = menuItems.indexOf(item);
            switch (position) {
                case 0:
                    string = R.string.menu_home;
                    drawable = R.drawable.ic_menu_home;
                    break;
                case 1:
                    string = R.string.menu_tag;
                    drawable = R.drawable.ic_menu_tag;
                    break;
                case 2:
                    string = R.string.menu_user;
                    drawable = R.drawable.ic_menu_user;
                    break;
                case 3:
                    string = R.string.menu_setting;
                    drawable = R.drawable.ic_menu_setting;
                    break;
            }
            item = initBottomNav(string, position, drawable);
            menuItems.set(position, item);
        }
        mBind.bottomNavigation.setOnItemSelectedListener(this);
    }

    private MenuItem initBottomNav(@StringRes int tag, int index, @DrawableRes int drawable) {
        mBind.bottomNavigation.getMenu().add(0, index, index, tag);
        mBind.bottomNavigation.getMenu().getItem(index).setIcon(drawable);
        return mBind.bottomNavigation.getMenu().getItem(index);
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        showFragment(item.getItemId());
        return true;
    }

    int mLastSelectedPosition = -1;
    private void showFragment(int position) {
        if(position != mLastSelectedPosition) {
            changeTitle(position);
            mBind.content.setCurrentItem(position, true);
            mLastSelectedPosition = position;
        }else {
            scrollToTop(position);
        }
    }

    private long lastClickTime;
    private void scrollToTop(int position) {
        long now = System.currentTimeMillis();
        if((now - lastClickTime) < 1200) {
            Fragment fragment = fragmentPager.createFragment(position);
            if(fragment instanceof BaseFreshFragment) {
                lastClickTime = 0;
                ((BaseFreshFragment<?>) fragment).scrollTo(0);
            }
            mBind.appBarLayout.setExpanded(true, true);
        }
        lastClickTime = now;
    }

    private void changeTitle(int position) {
        String title;
        switch (position) {
            case 0:
                title = getString(R.string.menu_home);
                mBind.animationView.setAnimation(R.raw.recommend);
                break;
            case 1:
                title = getString(R.string.menu_tag);
                mBind.animationView.setAnimation(R.raw.setting);
                break;
            case 2:
                title = getString(R.string.menu_user);
                mBind.animationView.setAnimation(R.raw.myself);
                break;
            case 3:
                title = getString(R.string.menu_setting);
                mBind.animationView.setAnimation(R.raw.search);
                break;
            default:
                title = "";
                break;
        }
        mBind.animationView.setRepeatCount(LottieDrawable.INFINITE);
        mBind.animationView.playAnimation();

        ObjectAnimator alpha = ObjectAnimator.ofFloat(mBind.tvTitle, "alpha", 1f, 0f);
        ObjectAnimator alpha2 = ObjectAnimator.ofFloat(mBind.tvTitle, "alpha", 0f, 1f);
        ObjectAnimator translationY = ObjectAnimator.ofFloat(mBind.tvTitle, "translationY"
                , mBind.tvTitle.getHeight(), 0f);
        ObjectAnimator translationY2 = ObjectAnimator.ofFloat(mBind.tvTitle, "translationY"
                , 0f, mBind.tvTitle.getHeight());
        AnimatorSet animSet = new AnimatorSet();
        long duration = 40L;
        alpha.setDuration(duration);
        alpha2.setDuration(duration);
        translationY.setDuration(duration);
        translationY2.setDuration(duration);

        alpha.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                mBind.tvTitle.setText(title);
            }
        });
        animSet.setDuration(duration * 4);
        animSet.play(alpha2).after(alpha);
        animSet.play(translationY).after(translationY2);
        animSet.start();
    }

}

File: app/src/main/java/com/lzm/lightLive/act/IntroActivity.kt
package com.lzm.lightLive.act

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.text.Editable
import android.util.*
import android.view.*
import android.widget.FrameLayout
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModel
import androidx.recyclerview.widget.GridLayoutManager
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.airbnb.lottie.LottieDrawable
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.navigation.NavigationBarView
import com.google.android.material.transition.platform.MaterialContainerTransformSharedElementCallback
import com.lzm.lib_base.BaseBindFragmentActivity
import com.lzm.lib_base.BaseFreshFragment
import com.lzm.lib_base.BaseScrollableFragment
import com.lzm.lib_base.behavior.BottomNavBarBehavior
import com.lzm.lightLive.R
import com.lzm.lightLive.adapter.FragmentAdapter
import com.lzm.lightLive.adapter.HostAdapter
import com.lzm.lightLive.databinding.ActivityIntroBinding
import com.lzm.lightLive.frag.AllCategoryFragment
import com.lzm.lightLive.frag.AllSubscribedFragment
import com.lzm.lightLive.frag.HomeFragment
import com.lzm.lightLive.frag.SettingFragment
import com.lzm.lightLive.http.BaseResult
import com.lzm.lightLive.http.RetrofitManager
import com.lzm.lightLive.http.bean.bl.BLRoom
import com.lzm.lightLive.http.request.bl.BlHttpRequest
import com.lzm.lightLive.model.RoomViewModel
import com.lzm.lightLive.util.CacheUtil
import com.lzm.lightLive.util.SimpleTextWatcher
import com.lzm.lightLive.util.UiTools
import com.lzm.lightLive.video.service.FloatingService
import com.lzm.lightLive.view.YouTuDraggingView
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlin.math.abs

class IntroActivity : BaseBindFragmentActivity<ActivityIntroBinding?, ViewModel?>(), NavigationBarView.OnItemSelectedListener {

    companion object {
        private const val TAG = "IntroActivity"

        var mPlayerBinder: FloatingService.PlayerBinder? = null

        private val mConnection: ServiceConnection = object : ServiceConnection {
            override fun onServiceConnected(name: ComponentName?, service: IBinder) {
                Log.e(TAG, "onServiceConnected: $name $service" )
                mPlayerBinder = service as FloatingService.PlayerBinder
            }

            override fun onServiceDisconnected(name: ComponentName?) {
                mPlayerBinder = null
            }
        }
    }

    private val itemHome: MenuItem? = null
    private val itemTag: MenuItem? = null
    private val itemUser: MenuItem? = null
    private val itemSearch: MenuItem? = null
    private val menuItems: MutableList<MenuItem?> = ArrayList()
    private var fragmentPager: FragmentAdapter? = null
    private var appBarSlideOffset = -1
    private var percent = 0f

    override fun getLayoutResId(): Int {
        return R.layout.activity_intro
    }

    override fun getViewModel(): ViewModel {
        return RoomViewModel()
    }

    override fun setViewModel(viewModel: ViewModel?) {
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        val window = window
        window.requestFeature(Window.FEATURE_ACTIVITY_TRANSITIONS)
        setExitSharedElementCallback(MaterialContainerTransformSharedElementCallback())
        window.sharedElementsUseOverlay = false
        super.onCreate(savedInstanceState)
        initView()
        initBottomNav()
        initFragment()
        initCustomSetting()
        test()
        mBind?.appBarLayout?.alpha = 1f
        mBind?.appBarLayout?.post { initAppBarListener() }
        showFragment(0)

//        val intent = Intent(this, FloatingService::class.java)
//        bindService(intent, mConnection, Context.BIND_AUTO_CREATE)
    }

    private fun test() {
        val mBlRequest = RetrofitManager.blRetrofit.create(
            BlHttpRequest::class.java
        )
        mBlRequest.getRoomInfo("5194110")
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(object : Observer<BaseResult<BLRoom?>> {
                override fun onSubscribe(d: Disposable) {}
                override fun onNext(result: BaseResult<BLRoom?>) {
                    Log.e(TAG, "getRoomInfo: ${result.data}" )

                    mBlRequest.getUidInfo(arrayOf(result.data?.uid))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(object : Observer<BaseResult<BLRoom?>> {
                            override fun onSubscribe(d: Disposable) {}
                            override fun onNext(result: BaseResult<BLRoom?>) {
                                Log.e(TAG, "onNext: ${result.data}" )
                            }
                            override fun onError(e: Throwable) {
                                Log.e(TAG, "onError: $e" )
                            }
                            override fun onComplete() { }
                        })
                }

                override fun onError(e: Throwable) {
                    Log.e(TAG, "getRoomInfo: " + e.message)
                }

                override fun onComplete() {}
            })

    }

    @SuppressLint("ClickableViewAccessibility")
    private fun initView() {
        mBind?.searchRv?.layoutManager = GridLayoutManager(this, 2)
    }

    private fun initCustomSetting() {
        try {
            val blurBehavior: BottomNavBarBehavior? =
                (mBind?.actionBarBlur?.layoutParams as CoordinatorLayout.LayoutParams).behavior as BottomNavBarBehavior?
            val navBehavior: BottomNavBarBehavior? =
                (mBind?.bottomNavigation?.layoutParams as CoordinatorLayout.LayoutParams).behavior as BottomNavBarBehavior?
            if (null != blurBehavior
                && null != navBehavior
            ) {
                blurBehavior.isIsNestScroll = CacheUtil.getNavBarIsNestScroll(this)
                navBehavior.isIsNestScroll = CacheUtil.getNavBarIsNestScroll(this)
            }
        } catch (e: Exception) {
            //ignore.
        }
    }

    fun ViewPager2.bindFragment(
        fm: FragmentManager,
        lifecycle: Lifecycle,
        fragments: List<Fragment>
    ): ViewPager2 {
        offscreenPageLimit = fragments.size - 1

        adapter = object : FragmentStateAdapter(fm, lifecycle) {
            override fun getItemCount(): Int = fragments.size
            override fun createFragment(position: Int): Fragment = fragments[position]
        }
        return this
    }

    private fun initFragment() {
        mBind?.content?.bindFragment(
            supportFragmentManager,
            this.lifecycle,
            listOf(HomeFragment(), AllCategoryFragment(), AllSubscribedFragment(), SettingFragment()),
        )
        val home = Pair<Int, Fragment>(0, HomeFragment())
        val cate = Pair<Int, Fragment>(1, AllCategoryFragment())
        val sub = Pair<Int, Fragment>(2, AllSubscribedFragment())
        val set = Pair<Int, Fragment>(3, SettingFragment())
        fragmentPager = FragmentAdapter(this@IntroActivity,
            home,
            cate,
            sub,
            set
        )
        mBind?.content?.offscreenPageLimit = 2
//        mBind?.content?.isUserInputEnabled = true
        mBind?.content?.adapter = fragmentPager
        mBind?.content?.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                mBind?.bottomNavigation?.selectedItemId = menuItems[position]!!.itemId
                super.onPageSelected(position)
            }
        })
    }

    private fun initAppBarListener() {
        val barLp = mBind?.actionBar?.layoutParams
        val tvTitleLp = mBind?.tvTitle?.layoutParams as FrameLayout.LayoutParams
        val barHeight = mBind?.actionBar!!.height
        val tvTitleTopMargin = tvTitleLp.topMargin
        val tvTitleLeftMargin = tvTitleLp.leftMargin //UiTools.dp2px(this, 30);
        val textSize = UiTools.pxToDp(this, mBind?.tvTitle!!.textSize)
        mBind?.fab?.setOnClickListener {
            if (mBind?.searchBar?.visibility == View.GONE) {
                mBind?.fab?.setImageResource(R.drawable.ic_close)
                UiTools.circleReveal(mBind?.searchBar!!, true)
            } else {
                mBind?.fab?.setImageResource(R.drawable.ic_search)
                UiTools.circleReveal(mBind?.searchBar!!, false)
                search("")
            }
            UiTools.circleReveal(mBind?.fab!!, true)
        }
        mBind?.searchBar?.addTextChangedListener(object : SimpleTextWatcher() {
            override fun afterTextChanged(s: Editable) {
                Log.e(TAG, "afterTextChanged: $s")
                search(s.toString())
            }
        })
        mBind?.appBarLayout?.addOnOffsetChangedListener { _: AppBarLayout?, verticalOffset: Int ->
            if (verticalOffset == appBarSlideOffset) {
                return@addOnOffsetChangedListener
            }
            appBarSlideOffset = verticalOffset
            percent = abs(verticalOffset * 1.0f) / mBind?.anchorView!!.top
            if (percent != 0f) {
                barLp?.height = (barHeight - barHeight * (0.6 * percent)).toInt()
                tvTitleLp.topMargin =
                    (tvTitleTopMargin - tvTitleTopMargin * (0.5 * percent)).toInt()
                tvTitleLp.leftMargin = (tvTitleLeftMargin + tvTitleLeftMargin * percent).toInt()
                runOnUiThread {
                    mBind?.actionBar?.layoutParams = barLp
                    mBind?.tvTitle?.layoutParams = tvTitleLp
                    mBind?.tvTitle?.textSize = (textSize - textSize * (0.3 * percent)).toFloat()
                }
            }
            if (abs(verticalOffset) >= mBind?.anchorView!!.top) { //滑动到顶了
                mBind?.actionBarBlur?.alpha = 1f
            } else {                                              //滑动中 percent = 收缩比例
                mBind?.actionBarBlur?.alpha = percent
            }
        }
    }

    private fun initBottomNav() {
        menuItems.add(itemHome)
        menuItems.add(itemTag)
        menuItems.add(itemUser)
        menuItems.add(itemSearch)
        for (item in menuItems) {
            @StringRes var string = -1
            @DrawableRes var drawable = -1
            val position = menuItems.indexOf(item)
            when (position) {
                0 -> {
                    string = R.string.menu_home
                    drawable = R.drawable.ic_menu_home
                }
                1 -> {
                    string = R.string.menu_tag
                    drawable = R.drawable.ic_menu_tag
                }
                2 -> {
                    string = R.string.menu_user
                    drawable = R.drawable.ic_menu_user
                }
                3 -> {
                    string = R.string.menu_setting
                    drawable = R.drawable.ic_menu_setting
                }
            }
            menuItems[position] = initBottomNav(string, position, drawable)
        }
        mBind?.bottomNavigation?.setOnItemSelectedListener(this)
    }

    private fun initBottomNav(
        @StringRes tag: Int,
        index: Int,
        @DrawableRes drawable: Int
    ): MenuItem {
        mBind?.bottomNavigation?.menu?.add(0, index, index, tag)
        mBind?.bottomNavigation?.menu?.getItem(index)?.setIcon(drawable)
        return mBind?.bottomNavigation?.menu!!.getItem(index)
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        showFragment(item.itemId)
        return true
    }

    private var mLastSelectedPosition = -1
    private fun showFragment(position: Int) {
        if (position != mLastSelectedPosition) {
            changeTitle(position)
            mBind?.content?.setCurrentItem(position, true)
            mLastSelectedPosition = position
        } else {
            scrollToTop(position)
        }
    }

    private var lastClickTime: Long = 0
    private fun scrollToTop(position: Int) {
        val now = System.currentTimeMillis()
        if (now - lastClickTime < 1200) {
            mBind?.appBarLayout?.setExpanded(true, true)
            val fragment = fragmentPager!!.createFragment(position)
            if (fragment is BaseFreshFragment<*, *>) {
                lastClickTime = 0
                fragment.scrollTo(0)
            } else if (fragment is BaseScrollableFragment<*>) {
                lastClickTime = 0
                fragment.scrollToTop()
            }
        }else {
            lastClickTime = now
        }
    }

    private fun changeTitle(position: Int) {
        val title: String
        when (position) {
            0 -> {
                title = getString(R.string.menu_home)
                mBind?.animationView?.setAnimation(R.raw.recommend)
            }
            1 -> {
                title = getString(R.string.menu_tag)
                mBind?.animationView?.setAnimation(R.raw.setting)
            }
            2 -> {
                title = getString(R.string.menu_user)
                mBind?.animationView?.setAnimation(R.raw.myself)
            }
            3 -> {
                title = getString(R.string.menu_setting)
                mBind?.animationView?.setAnimation(R.raw.search)
            }
            else -> title = ""
        }
        mBind?.animationView!!.repeatCount = LottieDrawable.INFINITE
        mBind?.animationView!!.playAnimation()
        val alpha = ObjectAnimator.ofFloat(mBind?.tvTitle, "alpha", 1f, 0f)
        val alpha2 = ObjectAnimator.ofFloat(mBind?.tvTitle, "alpha", 0f, 1f)
        val translationY = ObjectAnimator.ofFloat(
            mBind?.tvTitle, "translationY", mBind?.tvTitle!!.height.toFloat(), 0f
        )
        val translationY2 = ObjectAnimator.ofFloat(
            mBind?.tvTitle, "translationY", 0f, mBind?.tvTitle!!.height.toFloat()
        )
        val animSet = AnimatorSet()
        val duration = 40L
        alpha.duration = duration
        alpha2.duration = duration
        translationY.duration = duration
        translationY2.duration = duration
        alpha.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                super.onAnimationEnd(animation)
                mBind?.tvTitle?.text = title
            }
        })
        animSet.duration = duration * 4
        animSet.play(alpha2).after(alpha)
        animSet.play(translationY).after(translationY2)
        animSet.start()
    }

    private fun search(searchText: String) {
        val fragment = fragmentPager!!.createFragment(mLastSelectedPosition)
        if (fragment is HomeFragment) {
            val searchData = fragment.search(searchText)
            if (null != searchData) {
                UiTools.circleReveal(mBind?.searchRv!!, true)
                val mSearchAdapter = HostAdapter(R.layout.layout_item_host)
                mBind?.searchRv!!.adapter = mSearchAdapter
                mSearchAdapter.setNewData(searchData)
            } else {
                UiTools.circleReveal(mBind?.searchRv!!, false)
            }
        }
    }

    override fun onDestroy() {
        unbindService(mConnection)
        super.onDestroy()
    }
}